SWIPs/swip-message_to_honey.md

```diff
@@ -26,42 +25,56 @@ This SWIP is part of a series of SWIPs (but can be implemented on it's own). To
+### Technical details
+This section describes the interaction between the nodes and the oracle in more detail.
+* Querying the oracle will return a `messagePrices` object, specifying a `Time to Live (TTL)` in seconds and `prices` object objects. The `prices` object contains an entry for each messageType (`swarmMessage`) and `swarmMessage` maps a `validFrom` to a respective `price`. Taken together, the oracle returns: 
 ```typescript
-{honeyPrices: 
+{messagePrices: 
   {
     TTL, 
-    prices: [0: {validFrom, msgPrices[]}, n:{...}]
+    prices: {
+      swarmMessageX: {
+        validFrom0: price0, // at time validFrom0, price0 will be applied
+        ...,                // 
+        validFromN: priceN
+      }
+      ...,
+      swarmMessageY: {
+        validFrom0: price0,
+        ...,
+        validFromN: priceN
+      }
+    }
   }
 }
 ```
-* After `TTL` expires, nodes will query the price oracle for a new `honeyPrices` object. 
-* The `TTL` is set as a state-variable in the smart-contract and can be updated.
-* To ensure that it is not possible that a new entry in the `prices` array becomes valid before all nodes are updated, the smart-contract or update-policy must ensure that the `validFrom` of new prices must be at least `TTL` seconds in the future.
-* Upon start-up, nodes will look at the contents of their local cache and query the price oracle if no `honeyPrices` object is found, or the `TTL` of their cached `honeyPrices` object has expired. 
-* If the oracle cannot be reached when the `TTL` of the last `honeyPrices` object is expired, nodes will continue to apply the `prices` as instructed by the expired `honeyObject` and will start to query the oracle at a more regular interval to re-establish connection. 
+* The applied price for a message is the price where `validFrom` is in the most recent past.
+* After `TTL` expires, nodes will query the price oracle for a new `messagePrices` object. 
+* Upon start-up, nodes will look at the contents of their local cache. If no `messagePrices` object is found, or the `TTL` of their cached `messagePrices` object has expired, the msgOracle will be queried. 
+* The `TTL` is set as a variable inside the implementation of the oracle and can be updated.
+* To ensure that a new entry in the 'smarmMessage' object can't become valid before all nodes are updated, the smart-contract or update-policy must ensure that the `validFrom` of new prices must be at least `TTL` seconds in the future.
+* If the oracle cannot be reached when the `TTL` of the last `messagePrices` object is expired, nodes will continue to apply the `prices` as instructed by the expired `messagePrices` object and will start to query the oracle at a more regular interval to re-establish connection. 
 
```

Comment:
I think we should add the edge case that a node starts but the oracle can not be reached. In that case, either
* The node starts up nonetheless and assumes some bogus price (e.g. 1), so that it can interact with the rest of the network. This will eventually result in being kicked out by every other node. So it delays the important stuff.
* The node has to be shutdown as this would be regarded as fatal. I vote for this option but please discuss. Maybe there are more options

Reply:
I think the source-code be updated as well when the price of the oracle gets updated. This price is obviously not guaranteed to reflect the actual oracle price, as nodes might be in different releases, but it is a reasonable fallback when the oracle cannot be reached upon startup. As the whole oracle history is always reflected on-chain, nodes can retrospectively apply the correct price when they eventually connect with the oracle. 
What do you think?

Reply:
If I understand you correctly, and I understand that you mean that when the oracle gets updated, we release a new version of the code, then I think that defeats the purpose of the oracle, and we should not do that.

I think this should be a question for research.

Reply:
I think you don't understand me correctly then. All nodes will look at the oracle for the most up-to-date price, but in case the oracle is not reachable at startup it is pleasant to have a fallback value in place. This does not make the oracle obsolete, it merely reduces the damage when the oracle can't be reached the first time a node boots up Swarm.

Another option is that the node tells the peers with whom he is interacting that he can't reach the oracle and the peers answer him the prices which they are using themselves => these prices will be applied.

We have 4 options now:
- The node starts up nonetheless and assumes some bogus price (e.g. 1) so that it can interact with the rest of the network. This will eventually result in being kicked out by every other node. So it delays the important stuff.
- The bogus value is  replaced by a "default" value, which is the oracle price at the time of the release of Swarm which is run by the node. This releases the damage even more, but eventually, accounting differences will get severe and the node is kicked out. 
- The node requests from his peer the oracle price they use. It uses the price which is mentioned by at least a big majority (say: 80%) of the answers. If not more than 80% of the nodes mention the same price, the node has to shut down.
- The node has to be shutdown directly as this would be regarded as fatal.

Involving research is a good idea!

Reply:
> but are you still saying we do a new release every time the oracle updates the price? 

No. Just saying that whenever there is a new release, it must update the fallback value of all messages to be the most-recent value known at that time.